### ISSUE TITLE
Insight aggregation: Add query validation checks for aggregation filter panel

### ISSUE THREAD

**Issue by vovakulikov** [sourcegraph/sourcegraph#39955]
Blocked by https://github.com/sourcegraph/sourcegraph/issues/39954

Since we're not going to support aggregation for all queries we need to disable some aggregation types and sometimes all aggregation panel if query has non supportable filters

Currently we do not support aggregation if
- query has AND or OR operators
- author and date aggregations types do not support query without type commit or type diff

/cc @joelkw @felixfbecker @vovakulikov @unclejustin

**Comment by Joelkw** [comment]
From sync – we should add the drilldown checks as well (can we add a `repo:` to certain queries? Can we not?)

**Comment by chwarwick** [comment]
related https://github.com/sourcegraph/sourcegraph/issues/39929

**Comment by vovakulikov** [comment]
From sync — we can use regular search stream results to understand which aggregation types are available for a given query. 

@chwarwick, do you have a clear checklist that we could use against search results? Is it just if repo matches are in there, we could aggregate by repositories? If the file path is in there, then file path aggregation is available.

If yes, does it work the same for author aggregation?  (Should we use just sync query checks if its type:diff or type:commit author aggregation is available and not checking search results for it?)

**Comment by felixfbecker** [comment]
> we can use regular search stream results to understand which aggregation types are available for a given query.

Does that mean we need to block on the search to finish before determining what aggregations to allow? That sounds like a bad waterfall-requests situation.

Could we just do a client-side check for checking if the query is `type:commit`/`type:author` to determine if we can group by author and check for a capture group to determine if we can group by that? I _think_ file path and repo should always be possible.

**Comment by chwarwick** [comment]
> Could we just do a client-side check for checking if the query is `type:commit`/`type:author` to determine if we can group by author and check for a capture group to determine if we can group by that? I _think_ file path and repo should always be possible.

Yes I think we should be able to get it from the query, but we also discussed that since the results are coming back we could also check what types of matches are present.  I think the only other limiting keyword currently besides `type:commit/diff` would be `select:repo` in which case I don't think there is anything to aggregate.

**Comment by felixfbecker** [comment]
I think it's reasonable to always remove `select:` operators for aggregation. If you think about the equivalent SQL, the default in our search is essentially `SELECT *`, and when we run an aggregation, we change that to `SELECT COUNT(*)` (you need to have an aggregation function selected in SQL when using `GROUP BY` or the query will error). Therefor, if the user had specified `SELECT repo` instead of `SELECT *`, it's still reasonable IMO to replace that with `SELECT COUNT(*)` (which basically means just removing any `select:` operator).

**Comment by vovakulikov** [comment]
I don't have a strong opinion on `select:repo` filter. But I'm curious what @AlicjaSuska @Joelkw think about it. So basically, there are two options.
- We disable the aggregation chart for any query that has a `select:` filter because select already does some aggregation in the aggregation chart; that query will be just a bunch of bars with value 1 (which isn't so useful) 
- As Felix suggested above, for the aggregation chart, we run the query without `select:` filters (so we change the original user search query), which means the aggregation chart will have different values than the original search query results.

**Comment by Joelkw** [comment]
* For `select:repo` let's run it without the `select:` operator and show the repo aggregations. 
* For `select:file` let's run it *with* the `select:file` operator and show the repo aggregations (useful to see which repos have most files). It is not true that "any query that has a select filter means we get a chart with bar values all 1" (as described in this example, seeing which repos have the most file results). 

If we can't do both, prioritize the described behavior for `select:file` and just show nothing/1s in the repo case. 

Anyone strongly disagree with this plan? cc @felixfbecker

**Comment by felixfbecker** [comment]
@Joelkw that only talks about the proactive case, but keep in mind the user can freely choose the aggregation attribute, so we need to express this more in terms of which `select:` and grouping attribute pairs are compatible and which are not. What I'm reading from your proposal is essentially: if they are **different** attributes, keep the `select:` filter, if they are the **same**, remove it (because that _would_ create the case where the bars are all 1).

This is because (as I just realized) the `select:` operator is not directly equivalent to `SELECT` in SQL, it's actually equivalent to `SELECT DISTINCT` because it [removes duplicates](https://docs.sourcegraph.com/code_search/reference/language#select). @vovakulikov is therefore right in that the `select` operator does perform a kind of aggregation already (it is not a pure projection). @Joelkw's proposal is to allow the user to use that as a "pre-aggregation" before our aggregation, except where it would prevent us from doing any sensible aggregation because they are the same attribute and it doesn't make sense to aggregate on the same attribute twice.

That does sound reasonable at first to me, but I'd like to point out that **this is inconsistent with how SQL works** and breaks the analogy, especially if we want to expose a **`groupby:` query operator** later (and build on the similarity to SQL in our user education). In SQL, `GROUP BY` is performed first, `SELECT` last. The `SELECT` is something like `COUNT(*)` in our case, imagine we had `select:count` that would be what we're really selecting (we can consider that implied when we're grouping is used since we don't yet have other aggregation functions).

If we wanted to represent _"select distinct files first, then group by repo, then select the count"_, or expressed differently, _"group by files first, then group by repo, then select the count"_ I think that would be something similar to these options in the future:

1. **A subquery**
  SQL: `SELECT COUNT(*) FROM (SELECT DISTINCT file FROM myquery) GROUP BY repo`
  Sourcegraph: `groupby:repo (myquery select:file)`

2. **A subquery without `SELECT DISTINCT` but only `GROUP BY`**
  SQL: `SELECT COUNT(*) FROM (SELECT repo FROM myquery GROUP BY file) GROUP BY repo`
  Sourcegraph: `groupby:repo (myquery groupby:file)`

3. **A keyword in the aggregation function**
  SQL: `SELECT COUNT(DISTINCT file) FROM myquery`
  Sourcegraph: `groupby:repo select:count(file) myquery`

(note: in all cases the outer query has an implied `select:count`).

So we can either:
1. Go with @Joelkw's proposal to basically run `select:` first, then group and don't care about this future inconsistency with SQL, which could block other use cases in the future
2. Go with @Joelkw's proposal and try to maintain magic backward-compatible behavior in the future where sometimes `select:` runs first and sometimes `groupby:`
3. Go with @Joelkw's proposal to enable as much as possible now and change the behavior later when there is a better alternative to express it (since it's not part of the query language _yet_ but only UI buttons, it's not as strong of a "breaking change")
4. Block the case of using `select:` when grouping and show an error message when `select:` is used (because conceptually, we're selecting the `count`, the user can't select anything else – the use case @Joelkw describes would then be possible in the future once we have `groupby:`)
5. Strip `select:` consistently from the query (which prevents the use case @Joelkw described/will potentially not show the expected results in that case)

**Comment by Joelkw** [comment]
Thanks @felixfbecker for these great detailed thoughts. I agree that this is most similar to select distinct and in the future a separate "group by" would be useful. 

That said, I most agree with your option 3, because of how valuable we've seen this use case be for our existing users (just the select, without the additional grouping) and I'd be excited to see what people do with more functionality. 

The main/most valuable new use case(s) might be something like "I want to see how many places re-use this library" or "how many people have touched this library across different repos." (The first: select file, aggregate graph by repo; the second: select repo, aggregate by author, for a commit search). 

I think if we validate the uses here (by enabling this now via option 3), it also provides us user feedback and evidence for prioritizing the group by feature in the query language directly.

**Comment by vovakulikov** [comment]
Since this logic about aggregation types' availability will live on the backend, I close this one in favour of https://github.com/sourcegraph/sourcegraph/issues/40418
